### PR TITLE
feat: export/import project-package

### PR DESCRIPTION
**Pull request by jferic**
#### What this PR does / why we need it:
apis for export/import package of project

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](paste your link here)

#### Specified Reviewers:

/assign @luobily @chengjoey 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      Support export/import project package in DevOps platform        |
| 🇨🇳 中文    |        实现/优化了 DevOps 平台的项目工程包导入导出      |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: apistructs/project.go [hunk 1]
```diff
@@ -321,6 +345,21 @@ func (cc ResourceConfigsInfo) GetClusterName(workspace string) string {
...
+func (cc ResourceConfigsInfo) GetWSConfig(workspace string) *ResourceConfigInfo {
+	switch strings.ToLower(workspace) {
+	case "prod":
+		return cc.PROD
+	case "staging":
+		return cc.STAGING
+	case "test":
+		return cc.TEST
+	case "dev":
+		return cc.DEV
```
Comment (chengjoey) [inline apistructs/project.go:357]: could you use apistructs.DiceWorkspace instead of `prod` `test`...

File: bundle/addon.go [hunk 2]
```diff
@@ -289,3 +289,36 @@ func (b *Bundle) DeleteAddon(addonID string, orgID string, userID string) (*apis
 	}
 	return &resp.Data, nil
 }
+
+func (b *Bundle) CreateCustomAddon(userID, addonName, name string, orgId, projectId uint64, workspace string,
+	configs map[string]interface{}) (map[string]interface{}, error) {
```
Comment (chengjoey) [inline bundle/addon.go:294]: use struct instead of parameters

File: bundle/gittar.go [hunk 3]
```diff
@@ -686,3 +690,39 @@ func (b *Bundle) MergeRequestCount(userID string, req apistructs.MergeRequestCou
...
+	if err != nil {
+		return "", err
+	}
+
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return zipfile, err
+	}
+
+	return zipfile, nil
```
Comment (chengjoey) [inline bundle/gittar.go:727]: close file

File: bundle/release.go [hunk 4]
```diff
@@ -108,6 +112,76 @@ func (b *Bundle) ListReleases(req apistructs.ReleaseListRequest) (*apistructs.Re
...
+
+	f, err := os.Create(distDir + "/" + zipfile)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = io.Copy(f, resp.Body)
+	if err != nil {
+		return zipfile, err
+	}
```
Comment (chengjoey) [inline bundle/release.go:153]: close file

File: modules/dop/services/project/export_package.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+	envAddonInstances := map[string]map[string]apistructs.AddonFetchResponseData{
+		"dev":     {},
+		"test":    {},
+		"staging": {},
+		"prod":    {},
+	}
+	for _, a := range addonList.Data {
+		e := strings.ToLower(a.Workspace)
+		switch e {
+		case "dev", "test", "staging", "prod":
```
Comment (chengjoey) [inline modules/dop/services/project/export_package.go:173]: use apistructs.WorkSpace type is better

File: modules/dop/services/project/export_package.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+}
+
+func (t *PackageDB) SetMeta() error {
+	user, err := t.bdl.GetCurrentUser(t.IdentityInfo.UserID)
+	if err != nil {
+		return err
+	}
+	t.Package.MetaData.Creator = user.Nick
+
+	publicURL := os.Getenv("UI_PUBLIC_URL")
```
Comment (chengjoey) [inline modules/dop/services/project/export_package.go:381]: conf.UIPublicURL()

File: modules/openapi/api/apis/dop/project_package_import.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+var ProjectPackageImport = apis.ApiSpec{
+	Path:          "/api/orgs/<orgID>/projects/<projectID>/package/actions/import",
+	BackendPath:   "/api/orgs/<orgID>/projects/<projectID>/package/actions/import",
+	Host:          "dop.marathon.l4lb.thisdcos.directory:9527",
+	Scheme:        "http",
+	Method:        "POST",
+	TryCheckLogin: true,
+	RequestType:   apistructs.ImportProjectPackageRequest{},
+	Doc:           "summary: 导入项目模版",
```
Comment (chengjoey) [inline modules/openapi/api/apis/dop/project_package_import.go:30]: 项目工程包

File: bundle/addon.go (hunk 2 above)
Comment (jferic) [reply bundle/addon.go:294]: may be apistructs.CustomAddonCreateRequest, but it has too may feilds

File: modules/dop/endpoints/project.go [hunk 8]
```diff
@@ -681,3 +681,158 @@ func (e *Endpoints) ParseProjectTemplate(ctx context.Context, r *http.Request, v
...
+		return apierrors.ErrImportProjectPackage.InvalidParameter("orgID").ToResp(), nil
+	}
+	req.OrgID = orgID
+	req.IdentityInfo = identityInfo
+	if req.ProjectID == 0 {
+		return apierrors.ErrImportProjectPackage.InvalidParameter("projectID").ToResp(), nil
+	}
+	if req.OrgID == 0 {
+		return apierrors.ErrImportProjectPackage.InvalidParameter("orgID").ToResp(), nil
+	}
```
Comment (luobily) [inline modules/dop/endpoints/project.go:783]: You can add a func to get orgID / projectID / IdentityInfo.
Duplicate code in many places

File: modules/dop/endpoints/project.go [hunk 9]
```diff
@@ -681,3 +681,158 @@ func (e *Endpoints) ParseProjectTemplate(ctx context.Context, r *http.Request, v
...
+	exportReq.OrgID = orgID
+	// check permission
+	identityInfo, err := user.GetIdentityInfo(r)
+	if err != nil {
+		return apierrors.ErrExportProjectPackage.NotLogin().ToResp(), nil
+	}
+	exportReq.IdentityInfo = identityInfo
+	if exportReq.OrgID == 0 || exportReq.ProjectID == 0 {
+		return apierrors.ErrExportProjectPackage.InvalidParameter(fmt.Errorf("orgID and projectID can't be empty")).ToResp(), nil
+	}
```
Comment (luobily) [inline modules/dop/endpoints/project.go:705]: You can add a func to get orgID / projectID / IdentityInfo.
Duplicate code in many places

File: modules/dop/services/project/import_package.go [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+	ProjectID      uint64 `json:"projectID"`
+	ProjectName    string `json:"projectName"`
+	OrgID          int64  `json:"orgID"`
+	OrgName        string `json:"orgName"`
+	ProjectPackage *apistructs.ProjectPackage
+	bdl            *bundle.Bundle
+	packageName    string
+	apistructs.IdentityInfo
+	TmpDir string
+}
```
Comment (luobily) [inline modules/dop/services/project/import_package.go:49]: Duplicate struct in packageDB

File: apistructs/project.go [hunk 11]
```diff
@@ -503,3 +542,94 @@ type ProjectTemplateData struct {
...
+	Creator     string `yaml:"creator" json:"creator"`
+	Type        string `yaml:"type" json:"type"`
+	Source      string `yaml:"source" json:"source"`
+	Description string `yaml:"description,omitempty" json:"description,omitempty"`
+}
+
+type Artifact struct {
+	Type    string `yaml:"type" json:"type"`
+	Name    string `yaml:"name" json:"name"`
+	Version string `yaml:"version" json:"version"`
```
[the remark this answers is not in the document]
Comment (jferic) [reply apistructs/project.go:582]: Artifact for input paramters, ArtifactPkg for inner process

merge them to rm duplication

**Review by chengjoey — DISMISSED** [review]

File: bundle/addon.go (hunk 2 above)
Comment (luobily) [reply bundle/addon.go:294]: It is recommended to use apistructs.CustomAddonCreateRequest, which is better than writing so many params.

**Pull request merged**